panic: counter already signaled done. use a new counter. in addPeerToQuery

I've encountered this panic a few times in go-ipfs 0.4.19 on Windows 7 amd64 after a few hours of running:
```
panic: counter already signaled done. use a new counter.

goroutine 2351179 [running]:
gx/ipfs/QmQNQhNmY4STU1MURjH9vYEMpx2ncMS4gbwxXWtrEjzVAq/go-todocounter.(*todoCounter).Increment(0xc003d04540, 0xc000000001)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmQNQhNmY4STU1MURjH9vYEMpx2ncMS4gbwxXWtrEjzVAq/go-todocounter/counter.go:86 +0xb6
gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht.(*dhtQueryRunner).addPeerToQuery(0xc00d958580, 0xc0126e6450, 0x22)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht/query.go:200 +0x110
gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht.(*dhtQueryRunner).Run(0xc00d958580, 0x138ef40, 0xc0077b7f80, 0xc004d3fe90, 0x3, 0x3, 0x0, 0x0, 0x0)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht/query.go:136 +0x14a
gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht.(*dhtQuery).Run(0xc004d3fef0, 0x138ef40, 0xc0077b7f80, 0xc004d3fe90, 0x3, 0x3, 0x0, 0x0, 0x0)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht/query.go:70 +0x109
gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht.(*IpfsDHT).GetClosestPeers.func2(0xc0012c7800, 0xc00e867da0, 0xc004d3fef0, 0x138efc0, 0xc0015e7bc0, 0xc004d3fe90, 0x3, 0x3, 0xc008891a40, 0x22)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht/lookup.go:96 +0xe2
created by gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht.(*IpfsDHT).GetClosestPeers
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmdR6WN3TUEAVQ9KWE2UiFJikWTbUvgBJay6mjB4yUJebq/go-libp2p-kad-dht/lookup.go:92 +0x2fd
```

for reference, my daemon says:
```
go-ipfs version: 0.4.19-
Repo version: 7
System version: amd64/windows
Golang version: go1.11.5
```

0.4.19's package.json indicates go-libp2p-kad-dht version 4.4.31, so possibly this is related to #277? I haven't run recent earlier builds long enough to be sure this is a regression.

--- a/dial_queue.go
+++ b/dial_queue.go
@@ -4,7 +4,7 @@ import (
 	"context"
 	"fmt"
 	"math"
-	"sync/atomic"
+	"sync"
 	"time"
 
 	peer "github.com/libp2p/go-libp2p-peer"
@@ -32,9 +32,9 @@ const (
 type dialQueue struct {
 	*dqParams
 
-	nWorkers uint
-	out      *queue.ChanQueue
-	started  int32
+	nWorkers  uint
+	out       *queue.ChanQueue
+	startOnce sync.Once
 
 	waitingCh chan waitingCh
 	dieCh     chan struct{}
@@ -128,14 +128,13 @@ func newDialQueue(params *dqParams) (*dialQueue, error) {
 
 // Start initiates action on this dial queue. It should only be called once; subsequent calls are ignored.
 func (dq *dialQueue) Start() {
-	if !atomic.CompareAndSwapInt32(&dq.started, 0, 1) {
-		return
-	}
-	tgt := int(dq.dqParams.config.minParallelism)
-	for i := 0; i < tgt; i++ {
-		go dq.worker()
-	}
-	dq.nWorkers = uint(tgt)
+	dq.startOnce.Do(func() {
+		tgt := int(dq.dqParams.config.minParallelism)
+		for i := 0; i < tgt; i++ {
+			go dq.worker()
+		}
+		dq.nWorkers = uint(tgt)
+	})
 }
 
 func (dq *dialQueue) control() {

--- a/dial_queue.go
+++ b/dial_queue.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"math"
+	"sync/atomic"
 	"time"
 
 	peer "github.com/libp2p/go-libp2p-peer"
@@ -33,6 +34,7 @@ type dialQueue struct {
 
 	nWorkers uint
 	out      *queue.ChanQueue
+	started  int32
 
 	waitingCh chan waitingCh
 	dieCh     chan struct{}
@@ -90,9 +92,10 @@ type waitingCh struct {
 	ts time.Time
 }
 
-// newDialQueue returns an adaptive dial queue that spawns a dynamically sized set of goroutines to preemptively
-// stage dials for later handoff to the DHT protocol for RPC. It identifies backpressure on both ends (dial consumers
-// and dial producers), and takes compensating action by adjusting the worker pool.
+// newDialQueue returns an _unstarted_ adaptive dial queue that spawns a dynamically sized set of goroutines to
+// preemptively stage dials for later handoff to the DHT protocol for RPC. It identifies backpressure on both
+// ends (dial consumers and dial producers), and takes compensating action by adjusting the worker pool. To
+// activate the dial queue, call Start().
 //
 // Why? Dialing is expensive. It's orders of magnitude slower than running an RPC on an already-established
 // connection, as it requires establishing a TCP connection, multistream handshake, crypto handshake, mux handshake,
@@ -112,7 +115,6 @@ type waitingCh struct {
 func newDialQueue(params *dqParams) (*dialQueue, error) {
 	dq := &dialQueue{
 		dqParams:  params,
-		nWorkers:  params.config.minParallelism,
 		out:       queue.NewChanQueue(params.ctx, queue.NewXORDistancePQ(params.target)),
 		growCh:    make(chan struct{}, 1),
 		shrinkCh:  make(chan struct{}, 1),
@@ -120,11 +122,20 @@ func newDialQueue(params *dqParams) (*dialQueue, error) {
 		dieCh:     make(chan struct{}, params.config.maxParallelism),
 	}
 
-	for i := 0; i < int(params.config.minParallelism); i++ {
-		go dq.worker()
-	}
 	go dq.control()
 	return dq, nil
+}
+
+// Start initiates action on this dial queue. It should only be called once; subsequent calls are ignored.
+func (dq *dialQueue) Start() {
+	if !atomic.CompareAndSwapInt32(&dq.started, 0, 1) {
+		return
+	}
+	tgt := int(dq.dqParams.config.minParallelism)
+	for i := 0; i < tgt; i++ {
+		go dq.worker()
+	}
+	dq.nWorkers = uint(tgt)
 }
 
 func (dq *dialQueue) control() {

--- a/dial_queue_test.go
+++ b/dial_queue_test.go
@@ -42,6 +42,8 @@ func TestDialQueueGrowsOnSlowDials(t *testing.T) {
 		t.Error("unexpected error when constructing the dial queue", err)
 	}
 
+	dq.Start()
+
 	for i := 0; i < 4; i++ {
 		_ = dq.Consume()
 		time.Sleep(100 * time.Millisecond)
@@ -85,6 +87,8 @@ func TestDialQueueShrinksWithNoConsumers(t *testing.T) {
 	if err != nil {
 		t.Error("unexpected error when constructing the dial queue", err)
 	}
+
+	dq.Start()
 
 	// acquire 3 consumers, everytime we acquire a consumer, we will grow the pool because no dial job is completed
 	// and immediately returnable.
@@ -158,6 +162,8 @@ func TestDialQueueShrinksWithWhenIdle(t *testing.T) {
 		t.Error("unexpected error when constructing the dial queue", err)
 	}
 
+	dq.Start()
+
 	// keep up to speed with backlog by releasing the dial function every time we acquire a channel.
 	for i := 0; i < 13; i++ {
 		ch := dq.Consume()
@@ -209,6 +215,8 @@ func TestDialQueueMutePeriodHonored(t *testing.T) {
 	if err != nil {
 		t.Error("unexpected error when constructing the dial queue", err)
 	}
+
+	dq.Start()
 
 	// pick up three consumers.
 	for i := 0; i < 3; i++ {

--- a/query.go
+++ b/query.go
@@ -136,6 +136,11 @@ func (r *dhtQueryRunner) Run(ctx context.Context, peers []peer.ID) (*dhtQueryRes
 		r.addPeerToQuery(p)
 	}
 
+	// start the dial queue only after we've added the initial set of peers.
+	// this is to avoid race conditions that could cause the peersRemaining todoctr
+	// to be done too early if the initial dial fails before others make it into the queue.
+	r.peersDialed.Start()
+
 	// go do this thing.
 	// do it as a child proc to make sure Run exits
 	// ONLY AFTER spawn workers has exited.

